drgn/helpers/linux/list.py

```diff
@@ -277,3 +284,97 @@ def hlist_for_each_entry(
     type = head.prog_.type(type)
     for pos in hlist_for_each(head):
         yield container_of(pos, type, member)
+
+
+def llist_empty(head: Object) -> bool:
+    """
+    Return whether a llist is empty.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() == 0x0
+
+
+def llist_is_singular(head: Object) -> bool:
+    """
+    Return whether an llist has only one element.
+
+    :param head: ``struct llist_head *``
+    """
+    return head.first.value_() != 0x0 and head.first.next.value_() == 0x0
+
+
+def llist_first_entry(head: Object, type: Union[str, Type], member: str) -> Object:
+    """
+    Return the first entry in an llist.
+
+    The list is assumed to be non-empty.
+
+    See also :func:`llist_first_entry_or_null()`.
+
+    :param head: ``struct llist_head *``
+    :param type: Entry type.
+    :param member: Name of llist_node member in entry type.
+    :return: ``type *``
+    """
+    return container_of(head.first, type, member)
+
+
+def llist_first_entry_or_null(
+    head: Object, type: Union[str, Type], member: str
+) -> Object:
+    """
+    Return the first entry in an llist or ``NULL`` if the llist is empty.
+
+    See also :func:`llist_first_entry()`.
+
+    :param head: ``struct llist_head *``
+    :param type: Entry type.
+    :param member: Name of list_node member in entry type.
+    :return: ``type *``
+    """
+    if head.first.value_() == 0x0:
+        return NULL(head.prog_, head.prog_.pointer_type(head.prog_.type(type)))
+    else:
+        return container_of(head.first, type, member)
+
+
+def llist_next_entry(pos: Object, member: str) -> Object:
+    """
+    Return the next entry in an llist.
+
+    :param pos: ``type*``
+    :param member: Name of list_node member in entry type.
+    :return: ``type *``
+    """
+    return container_of(getattr(pos, member).next, pos.type_.type, member)
+
+
+def llist_for_each(head: Object) -> Iterator[Object]:
+    """
+    Iterate over all of the nodes in an llist.
+
+    :param head: ``struct llist_head *``
```

Comment:
It looks like this macro in the kernel source actually takes a `struct llist_node *`:

```c

/**
 * llist_for_each - iterate over some deleted entries of a lock-less list
 * @pos:	the &struct llist_node to use as a loop cursor
 * @node:	the first entry of deleted list entries
 *
 * In general, some entries of the lock-less list can be traversed
 * safely only after being deleted from list, so start with an entry
 * instead of list head.
 *
 * If being used on entries deleted from lock-less list directly, the
 * traverse order is from the newest to the oldest added entry.  If
 * you want to traverse from the oldest to the newest, you must
 * reverse the order by yourself before traversing.
 */
#define llist_for_each(pos, node)			\
	for ((pos) = (node); pos; (pos) = (pos)->next)
```

Obviously the rationale for this doesn't apply to drgn, but in general, drgn helpers mimic the kernel's calling conventions. So, maybe we should take a `struct llist_node *pos` instead of `struct llist_head *head` for consistency with the kernel. It would need to be called as `llist_for_each(head.first)` if you do have a `struct llist_head *`, which doesn't seem like a big deal.

Reply:
Sure, this makes sense. I have made these helpers consistent with respect to their kernel counter parts now. This would avoid confusion in using these helpers